### ISSUE TITLE
Centralize logic for git ref handling

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#7877]
**What would you like Renovate to be able to do?**

Have centralized logic for determining if a git reference belongs to a known host type (GitHub, GitLab, etc) and resolving/enriching it automatically. So that each manager doesn't need to do it individually.

**Did you already have any implementation ideas?**

Some managers already have added logic to look at a git ref and determine "this is github.com" before assigning it as `github-tags`. Others may simply ignore all git refs if they *don't* match github.com or known hosts. Similarly, we have demand to be able to resolve git refs on self-hosted platforms like GitHub and GitLab, so there's been logic added to check hostRules to determine if a host is a platform or not.

Ideally this should be centralized so that either:
- Each manager can call a shared function to resolve it, or
- The `git-refs` datasource could perhaps do this at lookup time and "enrich" the results the same way `github-tags` does with `github-releases` content

I'm not sure yet which approach is preferable. The first approach would mean that the lookup type is more visible/transparent throughout the flow.

**Comment by lukasmrtvy** [comment]
Would be nice to have compatibility also for new, potentional gitlab-releases datasource #7534

**Comment by nejch** [comment]
This would also open up using renovate for a variety of custom datasources that usually track commit hashes (not tags). This is often done in systems such as yocto & related build systems that build from source:

- CMake recommends using commit hashes for ExternalProject (https://github.com/Kitware/CMake/blob/master/Modules/ExternalProject.cmake#L271-L272)
- yocto/bitbake recipes (https://docs.yoctoproject.org/3.2/ref-manual/ref-variables.html#term-SRCREV)
- also used in isar (https://github.com/ilbers/isar/blob/master/doc/user_manual.md#add-a-custom-application)
- kas refspecs (https://kas.readthedocs.io/en/1.0/userguide.html#project-configuration)
- I'm sure there's a lot more and many package managers support tracking hashes as well (like pip)

**Comment by jbreu** [comment]
I agree with @nejch , such a capability would enable a great set of additional build / configuration systems. Supporting yocto/bitbake recipes would instantly enable renovate for a huge set of embedded projects in FOSS space and in the SW industry.

**Comment by rarkins** [comment]
We should encourage managers to use `git-tags` datasource and enhance this datasource to recognize common patterns and delegate to `github-tags`, etc.
